Add guidance for the use of collections of entity types (#482)
Microsoft/api-guidelines: graph/collection-properties.md

> OData services treat collections of complex types differently than they treat collections of entity types. 

Reviewer: Isn't this missing a title?

> OData services treat collections of complex types differently than they treat collections of entity types. 

Reviewer: Not sure we should start a topic with "Odata services treat"

```suggestion
Microsoft Graph treats collections of complex types differently than collections of entity types. 
```

> This is due to the nature of entity types being "individually addressable" (they have some key which uniquely identifies them within the collection) while complex types are not individually addressable. 

Reviewer: Is there anywhere in this topic were we want to say something about collections of primitive types?

> The result of this difference is that it is almost always preferable to use entity types for collections rather than complex types; please see the exceptions [here](#exceptions).

Reviewer: I think the style guide is normally about brevity rather than politeness...

```suggestion
As a result, it is almost always preferable to use entity types for collections rather than complex types. See [exceptions](#exceptions) for acceptable usage of complex types in collections.
```

Let's use the following CSDL as an example:

```xml

> <EntityType Name="interestingData">

Reviewer: Can we have something a little more realistic?
I think we can do better than using foo and bar everywhere...

> <EntityType Name="interestingData">

Reviewer: Is this intended to be a singleton?  We might want to mention that, as earlier we point out that entity types have keys :)

  <Property Name="foos" Type="Collection(self.foo)" Nullable="false" />
  <NavitationProperty Name="bars" Type="Collection(self.bar)" Nullable="false" ContainsTarget="true" />
</EntityType>

<ComplexType Name="foo">
  <Property Name="someProperty" Type="Edm.String" Nullable="false" />
</ComplexType>

<EntityType Name="bar">
  <Key>
    <PropertyRef Name="id" />
  </Key>
  <Property Name="id" Type="Edm.String" Nullable="false" />
  <Property Name="differentProperty" Type="Edm.Int32" Nullable="false" />
</EntityType>
```

## Retrieving the elements in a collection

The [OData](https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#_Toc31358935) [standard](https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#_Toc31358947) specifies `foos` and `bars` both can be retrieved using a `GET` request:
> OData services support requests for data via HTTP GET requests.
> 
> ...
> 
> OData services support querying collections of entities, complex type instances, and primitive values.

> ### Complex Type

Reviewer 2: As i understood your suggestion was to prefer collection of entity types when there is a need to update individual elements therefore i would remove references to complex types and show difference only when it's absolutely essential so that the pattern would be relatively short. IMO if you show differences for GET then the POST doesn't add much value. While PATCH is important.

```HTTP
GET  /interestingData/foos

200 OK
{
  "value": [
    {

>       "someProperty": "an original value"

Reviewer 3: is there a reason that your complex collection example starts with one value, and your entity collection sample starts with 2? it makes the sample payloads look more different than they are.

    }
  ]
}
```

### Entity Type

```HTTP
GET  /interestingData/bars

200 OK
{
  "value": [
    {
      "id": "firstBarId",
      "differentProperty": 10
    },
    {
      "id": "secondBarId",
      "differentProperty": -6914
    }
  ]
}
```

## Adding individual elements to a collection

For both `foos` and `bars`, the OData standard specifies that elements can be added to the collection using a `POST` request

> 1. [Complex Types](https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_UpdateaCollectionProperty):

Reviewer 3: The protocol supports this, but it's a bit asymmetric to support adding but not removing, so I'm not sure it's a pattern we want to encourage on ms graph. at least we should all out that it's an uncommon pattern due to this asymmetry.

```HTTP
GET  /interestingData/bars

200 OK
{
  "value": [
    {
      "id": "firstBarId",
      "differentProperty": 10
    },
    {
      "id": "secondBarId",
      "differentProperty": -6914
    }
  ]
}
```

## Updating individual elements in a collection

The [OData standard](https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_UpdateanEntity) specifies that clients can update individual elements of a collection of entity types using a `PATCH` request:
> To update an individual entity, the client makes a PATCH or PUT request to a URL that identifies the entity.

There is no way to do this for complex types because there is no way to address a particular instance of a complex type within a collection. 

### Entity Type

#### Update an element in the collection

```HTTP
PATCH  /interestingData/bars/firstBarId
{
  "differentProperty": "15"
}

204 No Content
```

#### Check the new contents of the collection

```HTTP
GET  /interestingData/bars

200 OK
{
  "value": [
    {
      "id": "firstBarId",
      "differentProperty": 15
    },
    {
      "id": "secondBarId",
      "differentProperty": -6914
    }
  ]
}
```

> 

Reviewer 3: also, you can have other objects reference an entity within a collection, which you cannot do with a complex type:

```xml
<EntityType Name="baz">
      <Property Name="Description" Type="Edm.String"/>
      <NaviationProperty Name="bar" Type="self.bar"/>
</EntityType>
```

```HTTP
GET interestingData/baz
200 OK
{
    "Description": "my favorite bar",
    "bar": 
    {
        "id": "firstBarId",
        "differentProperty": 15
     }
}
```

## Updating a collection

The [OData standard](https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_UpdateanEntity) specifies that clients can replace all elements of a collection of complex types or all elements of a collection of entity types using a `PATCH` request:
> Collection properties...provided in the payload corresponding to updatable properties MUST replace the value of the corresponding property in the entity or complex type.

> The standard also specifies that collections of entities can be updated in a relative fashion using the [delta syntax](https://docs.oasis-open.org/odata/odata-json-format/v4.01/odata-json-format-v4.01.html#_Toc38457777) in a `PATCH` request:

Reviewer 3: maybe move this to after the replace request and just before the delta example.  Also call out (importantly) that this is not possible with complex types.

> The standard also specifies that collections of entities can be updated in a relative fashion using the [delta syntax](https://docs.oasis-open.org/odata/odata-json-format/v4.01/odata-json-format-v4.01.html#_Toc38457777) in a `PATCH` request:

Reviewer 3: I'm not sure "relative fashion" will resonate as strongly with what the user is trying to do here. maybe something like "The standard allows updating a collection of entities, including adding, removing, or modifying individual existing entities in a single PATCH request. This is not possible with complex types because there is no way to identity them individually within the collection."

    }
  ]
}
```

### Entity Type

#### Replace the elements in a collection

```HTTP
PATCH /interestingData
{
  "bars": [
    {
      "id": "fourthBarId",
      "differentProperty": 20
    },
    {
      "id": "fifthBarId",
      "differentProperty": -99999
    }
  ]
}

204 No Content
```

#### Check the new contents of the collection

```HTTP
GET /interestingData/bars

200 OK
{
  "value": [
    {
      "id": "fourthBarId",
      "differentProperty": 20
    },
    {
      "id": "fifthBarId",
      "differentProperty": -99999
    }
  ]
}
```

#### Change the elements in a collection

```HTTP
PATCH /interestingData/bars
{
  "@context": "#$delta",
  "value": [
    {
      "id": "fifthBarId",
      "differentProperty": 1024
    },
    {
      "id": "sixthBarId",
      "differentProperty": 9801
    },
    {
      "id": "fourthBarId",
      "@removed": {}
    }
  ]
}

204 No Content
```

#### Check the new contents of the collection

```HTTP
GET /interestingData/bars

200 OK
{
  "value": [
    {
      "id": "fifthBarId",
      "differentProperty": 1024
    },
    {
      "id": "sixthBarId",
      "differentProperty": 9801
    }
  ]
}
```

## Exceptions

> It might not make sense for a collection property to be a collection of entities in cases where the collection must be viewed holistically; that is, cases where one element in the collection impacts or is coupled with another element in the collection. 

Reviewer 3: OData does define the concept of [ordered collections](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#sec_PositionalInserts), with the ability to insert within the ordered collection according to ordinal. We could choose to adopt that pattern in graph, which would help facilitate these types of scenarios on graph and not fall back to a collection of complex types.

The most common example of this would be a list of "tasks" that should be executed in order.
Inserting a single task into a collection only makes sense because the client knows the preceeding and proceeding tasks.
In cases like this, it might make sense for a collection to use a complex type instead of an entity type.
However, even in this case, it might make sense to use a collection of entity types anyway so that clients can individually address the tasks in the list, having URLs to individual tasks that can be referred to. 
In such a case, the workload should simply not support a delta `PATCH` request on this collection.

## Conclusion

Collections of entity types have several behaviors that are not available for collections of complex types:

> 1. Individual elements can be retrieved

Author: Note that the corollary of this is that items in the collection can be referenced by other navigation properties

2. Individual elements can be removed
3. Individual elements can be updated
4. Several elements within the collection may be added, removed, or updated in a single request

> 

Reviewer 3: and individual entities within a collection can be referenced, i.e., by other entities.

Reviewer: Yes - this came up recently in a Defender API review - they wanted to reference an item in another collection, but that collection used complex types :(

> Further, to remove or update elements in a collection of complex types, the entire collection must be replaced with a `PATCH` request. 

Reviewer 3: suggest: "within"

> This has the potential to result in data loss for clients who accidentally don't include all of the current elements of the collection, or encounter a race condition where two clients are attempting to `PATCH` the same collection.

Reviewer 3: here I would just use "update"; the key is that they are updating the collection, not what verb they are using.
